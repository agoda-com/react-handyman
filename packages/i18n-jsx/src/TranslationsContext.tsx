import * as React from 'react'

export interface Translations {
  [key: string]: string
}

const TranslationsContext = React.createContext<Translations>({})

export default TranslationsContext
