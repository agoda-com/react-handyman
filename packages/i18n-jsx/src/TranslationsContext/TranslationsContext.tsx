import * as React from 'react';
import { Translations } from './types';

const TranslationsContext = React.createContext<Translations>({});

export default TranslationsContext;
