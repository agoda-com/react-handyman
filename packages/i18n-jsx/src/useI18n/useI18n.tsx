import * as React from 'react';
import { useSelector, ReactReduxContext } from 'react-redux';

import { format } from 'format-to-jsx';
import I18nContext, { Translations } from '../I18nContext';
import handleNotFound from '../utils/handleNotFound';
import { I18nSelector } from './selector';

const makeI18n = (translations: Translations): I18nSelector => (k, notFound, ...args) =>
  format(translations[k] ? translations[k] : handleNotFound(k, notFound), ...args);

const useI18n = () => {
  const { selector } = React.useContext(I18nContext);
  const store = React.useContext(ReactReduxContext);
  if (!store) {
    return makeI18n({});
  }
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const translations = useSelector(selector) || {};
  return makeI18n(translations);
};

export default useI18n;
