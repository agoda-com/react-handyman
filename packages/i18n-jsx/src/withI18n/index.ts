import withI18n from './withI18n';

export default withI18n;
