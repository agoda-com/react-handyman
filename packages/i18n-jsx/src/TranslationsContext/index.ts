import TranslationsContext from './TranslationsContext'

export * from './types'
export default TranslationsContext
