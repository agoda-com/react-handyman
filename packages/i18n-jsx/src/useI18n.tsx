import * as React from 'react'
import TranslationsContext from './TranslationsContext'

export type I18nSelector = (k: string, def: string, ...args: (string | number)[]) => string

const format = (template: string, ...args: (string | number)[]): string => {
  return String(template).replace(/\{([^{}]+)\}/g, (match, key: any) => (args[key] != null ? args[key] : key))
}

const getDef = (def: any) => {
  if (Array.isArray(def)) {
    return def.reduce((acc, e) => {
      if (Number.isInteger(e)) return acc + `{${e}}`
      return acc + e
    }, '')
  }

  return def
}

const useI18n = () => {
  const translations = React.useContext(TranslationsContext)

  const i18n: I18nSelector = (k: string, def: string, ...args: (string | number)[]) => {
    return format(translations[k] ? translations[k] : getDef(def), ...args)
  }

  return i18n
}

export default useI18n
