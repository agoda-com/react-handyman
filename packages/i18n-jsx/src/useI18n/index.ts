import useI18n from './useI18n'
export * from './types'
export default useI18n
