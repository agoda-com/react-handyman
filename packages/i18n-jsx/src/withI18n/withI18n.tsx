import * as React from 'react'
import useI18n from '../useI18n/useI18n'
import { I18nSelector } from '../useI18n/selector'

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type Optionalize<T extends K, K> = Omit<T, keyof K>;

const withI18n = <TProps extends TI18nProps, TI18nProps extends {}>(
  Component: React.ComponentType<TProps>,
  mapI18nToProps: (i18n: I18nSelector) => TI18nProps
) => {
  type OwnProps = Optionalize<TProps, TI18nProps>;

  const Wrapped: React.FC<OwnProps> = React.memo(props => {
    const i18n = useI18n()
    return <Component {...props as TProps} {...mapI18nToProps(i18n)} />
  });

  Wrapped.displayName = `withI18n(${Component.displayName})`;
  return Wrapped;
}

export default withI18n;
