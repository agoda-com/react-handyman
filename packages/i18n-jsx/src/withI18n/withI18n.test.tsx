import 'jest';
import * as React from 'react';
import withI18n from './withI18n';
import { render } from '@testing-library/react'
import TranslationsContext from '../TranslationsContext'
import { I18nSelector } from '../useI18n'

interface Translations {
  translations: {
    name: string,
  },
}

interface BaseProps {}

describe('withI18n', () => {
  const mapI18nToProps = (i18n: I18nSelector): Translations => ({
    translations: {
      name: i18n(1, 'Some default value'),
    },
  });

  it('should render component with i18n translations', () => {

    const Component: React.FC<BaseProps & Translations> = (props) => <div>{props.translations.name}</div>;
    const ComponentWithTranslations = withI18n(Component, mapI18nToProps)

    const { asFragment } = render(
      <TranslationsContext.Provider value={{1: 'translatedName'}}>
        <ComponentWithTranslations />
      </TranslationsContext.Provider>
    )

    expect(asFragment()).toMatchSnapshot();
  });

});