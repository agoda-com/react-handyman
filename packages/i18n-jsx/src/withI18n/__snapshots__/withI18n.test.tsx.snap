// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withI18n should render component with i18n translations 1`] = `
<DocumentFragment>
  <div>
    translatedName
  </div>
</DocumentFragment>
`;
